File: flutter_app/lib/models/routine.dart
// lib/models/routine.dart
class Exercise {
  final String name;
  final int sets;
  final int reps;
  final int restSec;

  Exercise({
    required this.name,
    required this.sets,
    required this.reps,
    required this.restSec,
  });

  factory Exercise.fromJson(Map<String, dynamic> json) {
    return Exercise(
      name: json['name'] as String,
      sets: json['sets'] as int,
      reps: json['reps'] as int,
      restSec: json['restSec'] as int,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'name': name,
      'sets': sets,
      'reps': reps,
      'restSec': restSec,
    };
  }
}

class Routine {
  final String id;
  final String name;
  final List<Exercise> exercises;

  Routine({
    required this.id,
    required this.name,
    required this.exercises,
  });

  factory Routine.fromJson(Map<String, dynamic> json) {
    return Routine(
      id: json['_id'] as String,
      name: json['name'] as String,
      exercises: (json['exercises'] as List<dynamic>)
          .map((e) => Exercise.fromJson(e as Map<String, dynamic>))
          .toList(),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      '_id': id,
      'name': name,
      'exercises': exercises.map((e) => e.toJson()).toList(),
    };
  }
}

File: backend/src/models/Routine.ts
// src/models/Routine.ts
import { Schema, model, Document, Types } from 'mongoose';

export interface IExercise {
  name: string;
  sets: number;
  reps: number;
  restSec: number;
}

export interface IRoutine extends Document {
  userId: Types.ObjectId;
  name: string;
  exercises: IExercise[];
  createdAt: Date;
  updatedAt: Date;
}

const ExerciseSchema = new Schema<IExercise>(
  {
    name: { type: String, required: true },
    sets: { type: Number, required: true },
    reps: { type: Number, required: true },
    restSec: { type: Number, required: true },
  },
  { _id: false },
);

const RoutineSchema = new Schema<IRoutine>(
  {
    userId: { type: Schema.Types.ObjectId, ref: 'User', required: true, index: true },
    name: { type: String, required: true },
    exercises: { type: [ExerciseSchema], default: [] },
  },
  { timestamps: true },
);

export default model<IRoutine>('Routine', RoutineSchema);
